app/controllers/journeys/capitalallowances/structuresBuildingsAllowance/StructuresBuildingsAllowanceController.scala

```diff
@@ -0,0 +1,95 @@
+import navigation.CapitalAllowancesNavigator
+import pages.capitalallowances.structuresBuildingsAllowance.StructuresBuildingsAllowancePage
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import queries.Settable
+import services.SelfEmploymentService
+import services.SelfEmploymentService.clearDataFromUserAnswers
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import utils.Logging
+import views.html.journeys.capitalallowances.structuresBuildingsAllowance.StructuresBuildingsAllowanceView
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class StructuresBuildingsAllowanceController @Inject() (override val messagesApi: MessagesApi,
+                                                        navigator: CapitalAllowancesNavigator,
+                                                        identify: IdentifierAction,
+                                                        getData: DataRetrievalAction,
+                                                        requireData: DataRequiredAction,
+                                                        service: SelfEmploymentService,
+                                                        formProvider: StructuresBuildingsAllowanceFormProvider,
+                                                        val controllerComponents: MessagesControllerComponents,
+                                                        view: StructuresBuildingsAllowanceView)(implicit ec: ExecutionContext)
+    extends FrontendBaseController
+    with I18nSupport
+    with Logging {
+
+  def onPageLoad(taxYear: TaxYear, businessId: BusinessId, mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData) {
+    implicit request =>
+      val form = request.userAnswers
+        .get(StructuresBuildingsAllowancePage, businessId.some)
+        .fold(formProvider(request.userType))(formProvider(request.userType).fill)
+
+      Ok(view(form, mode, request.userType, taxYear, businessId))
+  }
+
+  def onSubmit(taxYear: TaxYear, businessId: BusinessId, mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData) async {
+    implicit request =>
+      formProvider(request.userType)
+        .bindFromRequest()
+        .fold(
+          formErrors => Future.successful(BadRequest(view(formErrors, mode, request.userType, taxYear, businessId))),
+          answer =>
+            for {
+              (editedUserAnswers, redirectMode) <- handleGatewayQuestion(answer, request, mode, businessId)
+              updatedUserAnswers                <- service.persistAnswer(businessId, editedUserAnswers, answer, StructuresBuildingsAllowancePage)
+            } yield Redirect(navigator.nextPage(StructuresBuildingsAllowancePage, redirectMode, updatedUserAnswers, taxYear, businessId))
+        )
+  }
+
+  private def handleGatewayQuestion(currentAnswer: Boolean,
```

Comment:
I recommend looking at ZeroEmissionGoodsVehicleController - it has better separation of concern and does not have this handleGatewayQuestion duplicates